### ISSUE TITLE
Numerical stability: Use log_softmax for cross-entropy

### ISSUE THREAD

**Issue by lewisnjue** [lewisnjue/nnetflow#29]
cross_entropy_loss currently computes softmax followed by log:\n\n    probability = logits.softmax(axis=-1)\n    ce_loss = - (targets * probability.log()).sum(axis=-1).mean()\n\nThis is less stable than using log_softmax, and can yield -inf when probabilities underflow.\n\nProposed change:\n\n    log_probs = logits.log_softmax(axis=-1)\n    ce_loss = - (targets * log_probs).sum(axis=-1).mean()\n\nBenefits:\n- Better numerical stability\n- Simpler backward (leverages existing log_softmax VJP)\n\nAdd tests that stress large/small logits.\n
